tests/serverpod_test_server/test_e2e/json_protocol_test.dart

```diff
@@ -0,0 +1,118 @@
+import 'dart:convert';
+
+import 'package:http/http.dart' as http;
+import 'package:serverpod_test_server/test_util/config.dart';
+import 'package:test/expect.dart';
+import 'package:test/scaffolding.dart';
+import 'package:web_socket_channel/web_socket_channel.dart';
+
+void main() {
+  group("Given a Serverpod server when fetching an object, ", () {
+    late http.Response response;
+
+    setUpAll(() async {
+      response = await http.post(
+        Uri.parse("${serverUrl}jsonProtocol"),
+        body: jsonEncode({"method": "getJsonForProtocol"}),
+      );
+    });
+
+    test('then it should return status code 200', () {
+      expect(
+        response.statusCode,
+        200,
+      );
+    });
+
+    test('then it should contain the "nested" key', () {
+      Map jsonMap = jsonDecode(response.body);
+      expect(
+        jsonMap,
+        contains('nested'),
+      );
+    });
+  });
+
+  group(
+      "Given a Serverpod server when fetching an object with server only field, ",
+      () {
+    late http.Response response;
+
+    setUpAll(() async {
+      response = await http.post(
+        Uri.parse("${serverUrl}jsonProtocol"),
+        body: jsonEncode({"method": "getJsonForProtocol"}),
+      );
+    });
+
+    test('then the nested object should not contain server-only field', () {
+      Map jsonMap = jsonDecode(response.body);
+      Map nestedMap = jsonMap['nested'];
+      expect(
+        nestedMap,
+        isNot(contains('serverOnlyScope')),
+      );
+    });
+  });
```

Comment:
Suggestion: Add first level validation that the field is not included.

As suggested in the endpoint, add a field for the first level as well and validate that it is not inlcuded.

If I look at the model, the `nested` field holds the same model again. To validate that the first level server only field is not included we should validate that `jsonMap['serverOnlyScope'];` does not exist.

It is good to validate the nested case as well.

Reply:
Make sure to validate the "first-level" and "nested" object for a streaming connection as well.